### PR TITLE
sys/shell: added UDP shell commands

### PR DESCRIPTION
**Pull request by haukepetersen**
~~based on #2719~~

This code is not compilable/runnable at the moment, as it depends on some IPv6 and UDP functionality, which is not yet merged...

Basically this code serves two purposes: (i) to have simple to use shell commands for sending and receiving UDP packets and (ii) for having a simple example on howto do this :-)

Update:
depends on #2430

### CODE REVIEW

File: sys/shell/commands/sc_udp.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    }
+    /* allocate UDP header */
+    udp = ng_udp_hdr_build(payload, NULL, 0, (uint8_t *)&port, 2);
+    if (udp == NULL) {
+        puts("Error: unable to allocate UDP header");
+        ng_pktbuf_release(payload);
+        return;
+    }
+    /* allocate IPv6 header */
+    ip = ng_ipv6_hdr_build(udp, NULL, 0, (uint8_t *)&ip_addr, sizeof(ip_addr));
```
Comment (authmillenon) [inline sys/shell/commands/sc_udp.c]: Would prefer use of `ng_netreg_hdr_build()` here, so we later can swap the IP header type easily depending on the format of the IP address.

File: sys/shell/commands/sc_udp.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    ng_pktsnip_t *payload, *udp, *ip;
+    ipv6_addr_t ip_addr;
+    ng_netreg_entry_t *sendto;
+
+    /* parse destination address
+     *
+     * For the future: we should read the type of address from the addr string,
+     * so figure out if IPv4 or IPv6 and build the header accordingly...
+     * So only IPv6 support for now */
+    if (_parse_addr(&ip_addr, addr) < 0) {
```
Comment (authmillenon) [inline sys/shell/commands/sc_udp.c]: We have `ng_ipv6_addr_from_str()` for that.

File: sys/shell/commands/sc_udp.c (hunk 1 above)
Comment (authmillenon) [inline sys/shell/commands/sc_udp.c]: (also: wouldn't require #2454 for this to compile, but only #2575)

File: sys/shell/commands/sc_udp.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /* allocate payload */
+    payload = ng_pktbuf_add(NULL, data, strlen(data), NG_NETTYPE_UNDEF);
+    if (payload == NULL) {
+        puts("Error: unable to copy data to packet buffer");
+        return;
+    }
+    /* allocate UDP header */
+    udp = ng_udp_hdr_build(payload, NULL, 0, (uint8_t *)&port, 2);
```
Comment (authmillenon) [inline sys/shell/commands/sc_udp.c]: Just for consistency and for sake of being an example I would also prefer `ng_netreg_hdr_build()` here.

File: sys/shell/commands/sc_udp.c (hunk 2 above)
Comment (haukepetersen) [inline sys/shell/commands/sc_udp.c]: Ha, that was the one I was looking for :-)

File: sys/shell/commands/sc_udp.c (hunk 3 above)
Comment (haukepetersen) [inline sys/shell/commands/sc_udp.c]: I disagree, why introduce an additional function call when the targeted function is clear at hand?

File: sys/shell/commands/sc_udp.c (hunk 1 above)
Comment (haukepetersen) [inline sys/shell/commands/sc_udp.c]: as long as we do not have support for IPv4, I do not really see the necessity of introducing an additional function call. Once IPv4 is introduced, some of this code has to be re-factored anyhow...

**Comment by haukepetersen** [comment]
made use of `ng_ipv6_addr_from_str()`

File: sys/shell/commands/shell_commands.c [hunk 4]
```diff
@@ -159,6 +159,10 @@ extern void _netif_send(int argc, char **argv);
 #endif
 #endif
 
+#ifdef MODULE_NG_UDP
+extern void _udp(int argc, char **argv);
```
Comment (authmillenon) [inline sys/shell/commands/shell_commands.c]: Just a heads-up: shell-handlers have the signature

```C
int void <name>(int argc, char **argv);
```

now ;-)

**Comment by OlegHahm** [comment]
As far as I can see this PR contains more than just adding the shell command, or am I mistaken?

For the shell command itself, I think this is typical case for https://github.com/RIOT-OS/RIOT/issues/2755.

**Comment by authmillenon** [comment]
Yes #2719 is still in here (which was merged). @haukepetersen please rebase to master.

File: sys/shell/commands/shell_commands.c (hunk 4 above)
Comment (haukepetersen) [inline sys/shell/commands/shell_commands.c]: yap, I saw that (though I have no idea why that was introduced...).

**Comment by haukepetersen** [comment]
will do.

**Comment by haukepetersen** [comment]
rebased. It still depends on #2430.

**Comment by authmillenon** [comment]
Come to think about it, why do are these commands automatically compiled in? They are more or less just test commands. Why aren't they just part of e.g. a UDP test application.

**Comment by authmillenon** [comment]
(if someone wants to use them in their application they can just copy it from there)

**Comment by haukepetersen** [comment]
in the end, everything related to the shell is more or less for testing... How about I change this PR into a UDP example application?

**Comment by authmillenon** [comment]
Yes and no. Most are for configuration or bootstrapping. I'd prefer the example application.

**Comment by OlegHahm** [comment]
btw that's what I meant with
> For the shell command itself, I think this is typical case for #2755.

**Comment by haukepetersen** [comment]
Ok. The only problem at the moment is, that we do not have a network stack bootstrapping/auto_init mechanism in place. So a sample application will look most ugly at the moment... I will spend some time on the init stuff today and see if I can finally come up with something.

**Comment by authmillenon** [comment]
The only thing still missing is device and MAC layer initialization. Everything above bootstraps already via their respective init functions called from `auto_init`

**Comment by haukepetersen** [comment]
Yap, but thats the interesting part :-) I will propose something here soon.

File: sys/shell/commands/sc_udp.c [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+#ifdef MODULE_NG_PKTDUMP
+static ng_netreg_entry_t _server = {NULL,
+                                    NG_NETREG_DEMUX_CTX_ALL,
+                                    KERNEL_PID_UNDEF};
+#endif
+
+
+void _send(char *addr, char *port, char *data)
+{
+    uint16_t port;
```
Comment (bapclenet) [inline sys/shell/commands/sc_udp.c]: Is that normal that port is redefined here?

Comment (haukepetersen) [inline sys/shell/commands/sc_udp.c]: propably not (This PR is not tested too well, and as we discussed above will be heavily re-structured anyway...)

**Comment by haukepetersen** [comment]
I close this in favor of #2895.

**Pull request closed without merging**